### ISSUE TITLE
Feature request: KeepSolid VPN Support

### ISSUE THREAD

**Issue by sinjinsmythe** [qdm12/gluetun#420]
Inclusion of Keepsolid VPN service into Gluetun.

KeepSolid utilises standard OpenVPN CRT files

Web Page for VPN is at https://www.vpnunlimitedapp.com/

Would be grateful if you could consider support for this service on your image.

**Comment by PrimeN** [comment]
I second this!

**Comment by qdm12** [comment]
Hey guys, sorry for the long delay.

How do you get the Openvpn configuration file?

It seems to be behind that login page: https://my.keepsolid.com/products/vpn/

Basically I need to have access to all their Openvpn configuration files to set it up in gluetun. Ideally if there is a public URL to a zip file or to each of their Openvpn configuration file, that would do.

Thanks for your help!

**Comment by sinjinsmythe** [comment]
> Hey guys, sorry for the long delay.
> 
> How do you get the Openvpn configuration file?
> 
> It seems to be behind that login page: https://my.keepsolid.com/products/vpn/
> 
> Basically I need to have access to all their Openvpn configuration files to set it up in gluetun. Ideally if there is a public URL to a zip file or to each of their Openvpn configuration file, that would do.
> 
> Thanks for your help!

I have the ovpn file here, it generates from your account when you select the server you want to set up a manual connection to.
Is there a way I can send this to you? I'm not sure what info i would need to strip out before posting to the public forum.  If you could tell me what to remove I'll post the configuration file here for you.
Thanks!

**Comment by qdm12** [comment]
Would you have easily access to all their Openvpn files (e.g. in a big zip file)? I ideally need them all to find all their hostnames so that I can list them in the code to have nice filters by country, region etc.

I usually skip on VPN providers not offering their Openvpn files publicly, since it's hard then to have a server information update mechanism, but I'll try to make an exception. Maybe I can find a way to update server information by bruteforcing DNS records (what is in place for Cyberghost for example, as they are also hiding things - for no reason really, but that's another story).

At least you can send a few (like 3-4) Openvpn files of different location and different network protocol, I might have a way to extrapolate other hostnames from them if they use for example country codes. You can send them to [quentin.mcgaw@gmail.com](mailto:quentin.mcgaw@gmail.com) to avoid leaking anything, just in case. You can remove the Openvpn username and password before sending, it's usually in an `auth.txt` file (check if the Openvpn config the line starting with `auth `). I will however need the CA cert, the static key v1 etc. (looking like [these](https://github.com/qdm12/gluetun/blob/master/internal/constants/ivpn.go#L11)) and other files bundled with the config.

**Comment by qdm12** [comment]
Hey guys, there is an image `qmcgaw/gluetun:vpnunlimited`. Wiki is updated, you might want to check the environment variables wiki page on how to set it up. I haven't tested it myself yet, but it *should* work.

Unfortunately, due to how VPN Unlimited operates, it requires you to bind mount (as a secret file or just as a normal file) your private key since this is what they use instead of the usual username+password combo.

Let me know if it works!

**Comment by sinjinsmythe** [comment]
Hey Quentin
Thanks so much for this update and all the effort at coding to set this up

, I have downloaded this tonight and started some testing to see if I can get it going. I'm a bit of a noob at this but determined to get it working, I have a lot of reading up to do on the WIKI.
I've attached screenshot of a log file of output when I'm trying to run it on my docker environment on a Synology NAS.  I ran the container with High Privilege and seemed to get past some permission issues relating to "WARN 💡 Tip: Are you passing NET_ADMIN capabilit" but now stuck on 
WARN TUN device is not available: open /dev/net/tun: no such file or directory.
![gluetun-vpnunlimited](https://user-images.githubusercontent.com/66390891/122481066-f1f7fc00-cfc5-11eb-90ff-4f2cb57dd89a.JPG)
Might not be able to get back to this until Sunday/Monday as so busy at work.
Thanks :-)

**Comment by qdm12** [comment]
Cool thanks!

`WARN TUN device is not available: open /dev/net/tun: no such file or directory.` should be fine, it creates it on the line just below it. And yes you should pass `--cap-add=NET_ADMIN` to the Docker run command as indicated in the readme.md. Let me know if you encounter difficulties setting it up in the coming days so I can improve documentation eventually. Cheers!

**Comment by sinjinsmythe** [comment]
> Cool thanks!
> 
> `WARN TUN device is not available: open /dev/net/tun: no such file or directory.` should be fine, it creates it on the line just below it. And yes you should pass `--cap-add=NET_ADMIN` to the Docker run command as indicated in the readme.md. Let me know if you encounter difficulties setting it up in the coming days so I can improve documentation eventually. Cheers!
> 
> Also if I have the time, I'll try it with one of the files you sent me since it should work with that. I'll keep you updated.

Thank you!

I noticed in my log files the provider defaulted to "pia" so I changed this over to "vpn unlimited" as shown in the environment variables, but it didn't seem to like the quotes.  I removed the quotes and it seems to have run but the provider is blank in the log file (attached) is this ok?

[_qmcgaw-gluetun1_logs.txt](https://github.com/qdm12/gluetun/files/6673719/_qmcgaw-gluetun1_logs.txt)

**Comment by qdm12** [comment]
So I tested it successfully... and there were a bunch of bugs, especially it was missing the client certificate file, which is specific to each server (ugh, like Cyberghost).

So now on top of the client.key (PRIVATE KEY), you also need to bind mount the certificate (see https://github.com/qdm12/gluetun/wiki/Client-certificate), and each server needs a different one. Not totally ideal sadly.

Anyway my command was:

```sh
docker run -it --rm --cap-add=NET_ADMIN -e VPNSP="vpn unlimited" -e COUNTRY=Belgium -v "$(pwd)/gluetun:/gluetun" qmcgaw/gluetun
```

With the right ./gluetun/client.crt and ./gluetun/client.key.

Unfortunately at this point, you might as well just use the [openvpn configuration directly](https://github.com/qdm12/gluetun/wiki/Openvpn-file), at the cost of:

- having to resolve hostnames yourself with `nslookup hostname.com` and replacing it in the config VS updating server information with new releases or [yourself](https://github.com/qdm12/gluetun/wiki/Updating-Servers)
- gluetun not knowing exactly what provider it's running and showing a bit less pretty logs

Let me know when you're happy with it though and I'll merge it in the main branch. No hurry either.

**Comment by sinjinsmythe** [comment]
I've been trying to test but getting an error 

`2021/06/20 16:03:14 ERROR files do not exist: /run/secrets/openvpn_clientkey and /gluetun/client.key
2021/06/20 16:03:14 INFO Shutdown successful
failed to resize tty, using default size`

I put my client.key and client.crt for the Romanian server into /configs/gluetun folder on my synology NAS drive but I dont think I'm referencing it correctly?  Sorry I'm trying to learn as I go along....

I'm trying to launch with the command below 

` docker run -it --rm --cap-add=NET_ADMIN -e VPNSP="vpn unlimited" -e COUNTRY=Romania -e TZ=europe/london -e PUID=1024 -e PGID=101 -v "$(pwd)/gluetun:/configs/gluetun" qmcgaw/gluetun:vpnunlimited`

as when i tried to run it with your previous command above, it tries to launch with gluetun:latest and gives the error

`2021/06/20 15:59:35 ERROR environment variable "VPNSP" value is "vpn unlimited" and can only be one of: cyberghost, fastestvpn, hidemyass, ivpn, mullvad, nordvpn, privado, pia, private internet access, privatevpn, protonvpn, purevpn, surfshark, torguard, vyprvpn, windscribe`

**Comment by qdm12** [comment]
Change `$(pwd)/gluetun:/configs/gluetun` to `$(pwd)/gluetun:/gluetun`. Note you can also use absolute paths instead of `pwd` if you prefer. Make sure you have both client.key and client.crt inside gluetun/ on your host as well.

> as when i tried to run it with your previous command above, it tries to launch with gluetun:latest and gives the error

That might had been a fluke from my side, pushing the wrong image. Just use `qmcgaw/gluetun` without the `:vpnunlimited` tag and it should be working (wait for the publish to be a green mark [here](https://github.com/qdm12/gluetun/actions/runs/954787274) though as it's still building on Github). Going out I'll be back in a few hours.

**Comment by sinjinsmythe** [comment]
Thanks! I think I may have figured it out now with lots of tinkering, just waiting on the green tick on publish now to get the updated provider list and will update you later.

**Comment by qdm12** [comment]
It's up now, I built and pushed the images manually as it seems Docker's github action workflow is sort of broken. I'm still tinkering with the automated build system... 😞

**Comment by sinjinsmythe** [comment]
I pulled the latest image and tried to run again, I think it may be trying to access ivpn servers rather that vpnunlimited somehow according to logs when I specify the Romanian torrent server.

`=========================================,
================ Gluetun ================,
=========================================,
==== A mix of OpenVPN, DNS over TLS, ====,
======= Shadowsocks and HTTP proxy ======,
========= all glued up with Go ==========,
=========================================,
=========== For tunneling to ============,
======== your favorite VPN server =======,
=========================================,
=== Made with ❤️ by github.com/qdm12 ====,
=========================================,
,
Running version unknown built on an unknown date (commit unknown),
,
,
🔧 Need help? https://github.com/qdm12/gluetun/issues/new,
💻 Email? quentin.mcgaw@gmail.com,
☕ Slack? Join from the Slack button on Github,
💰 Help me? https://github.com/sponsors/qdm12,
2021/06/20 23:22:22 INFO Alpine version: 3.13.5,
2021/06/20 23:22:22 INFO OpenVPN 2.4 version: 2.4.11,
2021/06/20 23:22:22 INFO OpenVPN 2.5 version: 2.5.2,
2021/06/20 23:22:22 INFO Unbound version: 1.13.0,
2021/06/20 23:22:22 INFO IPtables version: v1.8.6,
2021/06/20 23:22:22 ERROR cannot read Openvpn settings: environment variable "SERVER_HOSTNAME": invalid values found: value "ro.vpnunlimitedapp.com" at position 1: possible values are: at.gw.ivpn.net, au-nsw.gw.ivpn.net, be.gw.ivpn.net, bg.gw.ivpn.net, br.gw.ivpn.net, ca-qc.gw.ivpn.net, ca.gw.ivpn.net, ch.gw.ivpn.net, cz.gw.ivpn.net, de.gw.ivpn.net, dk.gw.ivpn.net, es.gw.ivpn.net, fi.gw.ivpn.net, fr.gw.ivpn.net, gb-man.gw.ivpn.net, gb.gw.ivpn.net, hk.gw.ivpn.net, hu.gw.ivpn.net, il.gw.ivpn.net, is.gw.ivpn.net, it.gw.ivpn.net, jp.gw.ivpn.net, lu.gw.ivpn.net, nl.gw.ivpn.net, no.gw.ivpn.net, pl.gw.ivpn.net, pt.gw.ivpn.net, ro.gw.ivpn.net, rs.gw.ivpn.net, se.gw.ivpn.net, sg.gw.ivpn.net, sk.gw.ivpn.net, ua.gw.ivpn.net, us-az.gw.ivpn.net, us-ca.gw.ivpn.net, us-dc.gw.ivpn.net, us-fl.gw.ivpn.net, us-ga.gw.ivpn.net, us-il.gw.ivpn.net, us-nj.gw.ivpn.net, us-nv.gw.ivpn.net, us-ny.gw.ivpn.net, us-tx.gw.ivpn.net, us-ut.gw.ivpn.net, us-wa.gw.ivpn.net,
2021/06/20 23:22:22 INFO Shutdown successful,`

This is the command im using to launch, ive taken out the username and passwords

`docker run -d --name gluetun --cap-add=NET_ADMIN -e VPNSP="vpn unlimited" -e COUNTRY=Romania -e SERVER_HOSTNAME=ro.vpnunlimitedapp.com -e TZ=europe/london -e PUID=1024 -e PGID=101 -v "/volume1/configs/gluetun:/gluetun" -e OPENVPN_CLIENTCRT_SECRETFILE=/gluetun/client.crt -e OPENVPN_CLIENTKEY_SECRETFILE=/gluetun/client.key -e OPENVPN_USER=redacted -e OPENVPN_PASSWORD=redacted qmcgaw/gluetun`

Will have another look tomorrow, thanks :-)

**Comment by sinjinsmythe** [comment]
Tried again this morning with the latest image pull but still getting the same references to the ivpn.net servers.

I wonder if it has anything to do with the recent commit a few days ago where ivpn configuration name was fixed to vpn unlimited?
![1624256047946](https://user-images.githubusercontent.com/66390891/122715718-3652f880-d261-11eb-906a-2bc28c61eb62.jpg)

**Comment by qdm12** [comment]
Sorry for the delay. It's fixed now, it was a bug in the code (shameful copy paste), sorry about that.

**Comment by sinjinsmythe** [comment]
Great job I'll test when I get home later today 😁

**Comment by sinjinsmythe** [comment]
Bingo!
Up and running with a Romanian IP address within my container! Now the fun begins! 
Thank you so much for all your help developing this setup. If you need any help with your documentation etc let me know. 😁😁😁

**Comment by qdm12** [comment]
Great, enjoy! Feel free to ask questions on the Github Discussions if you need any more help for anything.
